File: WU/internal/cli/options.go
package cli

import (
	"fmt"
	"os"

	"WU/internal/support"
)

type CLIOptions struct {
	TenantID     string
	ClientID     string
	ClientSecret string
	ProductID    string
	SubmissionID string

	SelectAll bool
	DryRun    bool
	OutPath   string

	Destination string
	Name        string

	GoLiveImmediate bool
	GoLiveDate      string

	VisibleToAccounts []int

	AutoInstallDuringOSUpgrade     bool
	AutoInstallOnApplicableSystems bool

	IsDisclosureRestricted bool
	PublishToWindows10s    bool

	MsContact             string
	ValidationsPerformed  string
	AffectedOems          []string
	IsRebootRequired      bool
	IsCoEngineered        bool
	IsForUnreleasedHardware bool
	HasUiSoftware         bool
	BusinessJustification string

	Chids []string

	NoUI       bool
	OfferFilter bool
}

func defaultCLIOptions() *CLIOptions {
	return &CLIOptions{
		Destination: "windowsUpdate",
		GoLiveImmediate: true,
		VisibleToAccounts: []int{},

		AutoInstallDuringOSUpgrade:     true,
		AutoInstallOnApplicableSystems: true,

		MsContact:            "feizh@microsoft.com",
		ValidationsPerformed: "Product assurance team full range tested",
		AffectedOems:         []string{"Lenovo"},
		IsRebootRequired:     false,
		IsCoEngineered:       false,
		IsForUnreleasedHardware: false,
		HasUiSoftware:        false,
		BusinessJustification: "to meet MDA requirements",

		Chids:       []string{},
		OfferFilter: true,
	}
}

func ParseCLIOptions(argv []string) (*CLIOptions, error) {
	o := defaultCLIOptions()
	m := ParseArgs(argv)

	o.TenantID = m.GetSingle("--tenant-id")
	o.ClientID = m.GetSingle("--client-id")
	o.ClientSecret = m.GetSingle("--client-secret")
	o.ProductID = m.GetSingle("--product-id")
	o.SubmissionID = m.GetSingle("--submission-id")

	o.SelectAll = m.HasFlag("--select-all")
	o.DryRun = m.HasFlag("--dry-run")
	o.OutPath = support.FirstNonEmpty(m.GetSingle("--out"), "shippinglabel.request.json")

	o.Destination = support.FirstNonEmpty(m.GetSingle("--destination"), o.Destination)
	o.Name = m.GetSingle("--name")

	o.GoLiveImmediate = !m.HasFlag("--schedule-go-live")
	if v := m.GetSingle("--go-live-date"); !support.IsBlank(v) {
		o.GoLiveDate = v
		o.GoLiveImmediate = false
	}

	for _, s := range m.GetMany("--visible-to-accounts") {
		n, err := support.ParseIntStrict(s)
		if err != nil {
			return nil, support.NewAPIError("--visible-to-accounts 需要整数，但输入为: " + s)
		}
		o.VisibleToAccounts = append(o.VisibleToAccounts, n)
	}

	if m.HasFlag("--auto-install-os-upgrade") {
		o.AutoInstallDuringOSUpgrade = true
	}
	if m.HasFlag("--no-auto-install-os-upgrade") {
		o.AutoInstallDuringOSUpgrade = false
	}
	if m.HasFlag("--auto-install-applicable") {
		o.AutoInstallOnApplicableSystems = true
	}
	if m.HasFlag("--no-auto-install-applicable") {
		o.AutoInstallOnApplicableSystems = false
	}

	o.IsDisclosureRestricted = m.HasFlag("--is-disclosure-restricted")
	o.PublishToWindows10s = m.HasFlag("--publish-to-windows10s")

	if v := m.GetSingle("--ms-contact"); !support.IsBlank(v) {
		o.MsContact = v
	}
	if v := m.GetSingle("--validations-performed"); !support.IsBlank(v) {
		o.ValidationsPerformed = v
	}

	if a := m.GetMany("--affected-oems"); len(a) > 0 {
		o.AffectedOems = append([]string{}, a...)
	}

	o.IsRebootRequired = m.HasFlag("--is-reboot-required")
	o.IsCoEngineered = m.HasFlag("--is-co-engineered")
	o.IsForUnreleasedHardware = m.HasFlag("--is-for-unreleased-hardware")
	o.HasUiSoftware = m.HasFlag("--has-ui-software")

	if v := m.GetSingle("--business-justification"); !support.IsBlank(v) {
		o.BusinessJustification = v
	}

	o.Chids = append([]string{}, m.GetMany("--chids")...)

	o.NoUI = m.HasFlag("--no-ui")
	if m.HasFlag("--no-filter") {
		o.OfferFilter = false
	}

	return o, nil
}

func PrintErr(err error) {
	fmt.Fprintln(os.Stderr, "\n❌ 错误：")
	fmt.Fprintln(os.Stderr, err.Error())
}

File: WU-npm/src/cli/options.ts
import { parseArgs, type ArgSet } from './args.js'
import type { WuConfig } from '../config/store.js'
import { firstNonEmpty, isBlank } from '../support/strings.js'
import { APIError } from '../support/errors.js'

export interface CLIOptions {
  tenantId: string
  clientId: string
  clientSecret: string
  productId: string
  submissionId: string
  selectAll: boolean
  dryRun: boolean
  outPath: string
  destination: string
  name: string
  goLiveImmediate: boolean
  goLiveDate: string
  visibleToAccounts: number[]
  autoInstallDuringOSUpgrade: boolean
  autoInstallOnApplicableSystems: boolean
  isDisclosureRestricted: boolean
  publishToWindows10s: boolean
  msContact: string
  validationsPerformed: string
  affectedOems: string[]
  isRebootRequired: boolean
  isCoEngineered: boolean
  isForUnreleasedHardware: boolean
  hasUiSoftware: boolean
  businessJustification: string
  chids: string[]
  noUi: boolean
  offerFilter: boolean
}

function parseIntStrict(s: string): number {
  if (!/^[+-]?\d+$/.test(s.trim())) throw new APIError('--visible-to-accounts 需要整数，但输入为: ' + s)
  return parseInt(s, 10)
}

export function assembleOptions(config: WuConfig, argv: string[]): CLIOptions {
  const m: ArgSet = parseArgs(argv)

  const goLiveDate = m.getSingle('--go-live-date')
  let goLiveImmediate = config.goLiveImmediate && !m.hasFlag('--schedule-go-live')
  if (!isBlank(goLiveDate)) goLiveImmediate = false

  let autoOsUpgrade = config.autoInstallDuringOSUpgrade
  if (m.hasFlag('--auto-install-os-upgrade')) autoOsUpgrade = true
  if (m.hasFlag('--no-auto-install-os-upgrade')) autoOsUpgrade = false

  let autoApplicable = config.autoInstallOnApplicableSystems
  if (m.hasFlag('--auto-install-applicable')) autoApplicable = true
  if (m.hasFlag('--no-auto-install-applicable')) autoApplicable = false

  const vtaCli = m.getMany('--visible-to-accounts')
  const visibleToAccounts = vtaCli.length > 0 ? vtaCli.map(parseIntStrict) : [...config.visibleToAccounts]

  const oemsCli = m.getMany('--affected-oems')
  const affectedOems = oemsCli.length > 0 ? [...oemsCli] : [...config.affectedOems]

  return {
    tenantId: m.getSingle('--tenant-id'),
    clientId: m.getSingle('--client-id'),
    clientSecret: m.getSingle('--client-secret'),
    productId: m.getSingle('--product-id'),
    submissionId: m.getSingle('--submission-id'),
    selectAll: m.hasFlag('--select-all'),
    dryRun: m.hasFlag('--dry-run'),
    outPath: firstNonEmpty(m.getSingle('--out'), 'shippinglabel.request.json'),
    destination: firstNonEmpty(m.getSingle('--destination'), config.destination),
    name: m.getSingle('--name'),
    goLiveImmediate,
    goLiveDate: isBlank(goLiveDate) ? '' : goLiveDate,
    visibleToAccounts,
    autoInstallDuringOSUpgrade: autoOsUpgrade,
    autoInstallOnApplicableSystems: autoApplicable,
    isDisclosureRestricted: m.hasFlag('--is-disclosure-restricted') || config.isDisclosureRestricted,
    publishToWindows10s: m.hasFlag('--publish-to-windows10s') || config.publishToWindows10s,
    msContact: firstNonEmpty(m.getSingle('--ms-contact'), config.msContact),
    validationsPerformed: firstNonEmpty(m.getSingle('--validations-performed'), config.validationsPerformed),
    affectedOems,
    isRebootRequired: m.hasFlag('--is-reboot-required') || config.isRebootRequired,
    isCoEngineered: m.hasFlag('--is-co-engineered') || config.isCoEngineered,
    isForUnreleasedHardware: m.hasFlag('--is-for-unreleased-hardware') || config.isForUnreleasedHardware,
    hasUiSoftware: m.hasFlag('--has-ui-software') || config.hasUiSoftware,
    businessJustification: firstNonEmpty(m.getSingle('--business-justification'), config.businessJustification),
    chids: [...m.getMany('--chids')],
    noUi: m.hasFlag('--no-ui'),
    offerFilter: !m.hasFlag('--no-filter'),
  }
}
